### ISSUE TITLE
npm: azure-iot-device-mqtt / amqp defect functions

### ISSUE THREAD

**Issue by encrypted90** [Azure/azure-iot-sdks#365]
Hello

Using the [Getting started with azure-iot-device-mqtt](https://www.npmjs.com/package/azure-iot-device-mqtt) code, the following error is raised:

```
this.client.on('error', errCallback);
               ^
TypeError: Cannot read property 'on' of undefined
```

Then using the [Getting started with azure-iot-device-amqp](https://www.npmjs.com/package/azure-iot-device-amqp) code, the following error is raised:

```
throw new Error('Must connect before creating links');
    ^
Error: Must connect before creating links
```

Is there a solution for any of these errors?

Greetings

**Comment by pierreca** [comment]
Hi @encrypted90 
as stated in the other issue, those docs are not up to date. I would suggest you look directly at the sample in our repository that illustrates how to use the client object.
- sample code: https://github.com/Azure/azure-iot-sdks/blob/master/node/device/samples/simple_sample_device.js
- how to run the sample: : https://github.com/Azure/azure-iot-sdks/blob/master/doc/get_started/node-run-sample.md

Apologies for the outdated doc issues. we're working on getting them updated.

**Comment by pierreca** [comment]
@encrypted90 to be more precise, as described in the sample, the following steps need to happen in order:
`client.fromConnectionString()` -> to instantiate the client
`client.open()` -> to create the connection
`client.on('...')` or `client.sendEvent()` -> to use the client over the previously opened connection.

the sample illustrates that. please let us know if this ended up working for you or if you're still running into issues.

**Comment by encrypted90** [comment]
Using the Amqp script I receive messages send by the device explorer. But I don't receive messages send by the Amqp script, or messages send by the C# getting started guide.

**Comment by pierreca** [comment]
Hi @encrypted90 
Let me rephrase to make sure I understand what you're saying (can  you please confirm each of those statements: the more details you add, the better)

- I'm assuming you're talking about `node/device/samples/simple_sample_device.js` ? 
- when you start the sample, you can see the client connect
- the sample starts listening and sending messages (should be pretty clear on the console output)
- you do see the `MessageEnqueued` result message for the messages the sample is sending to the cloud in the console output
- You don't see anything in the 'Data` tab of the Device Explorer when you click the `Monitor` button (having made sure the right device is selected in the drop down list)?
- When you use the `Message To Device` tab in the Device Explorer to send a message, you can actually see your messages sent on the console output of the sample

Did I get all that right? Pasting the console output and screenshots of the device explorer may help us figure out what's wrong. Thanks!

**Comment by encrypted90** [comment]
Hello Pierreca, regarding your questions:

1. Correct
2. Yes, client connects
3. Correct:

![console](https://cloud.githubusercontent.com/assets/4939067/13595312/8a11778e-e50a-11e5-8295-fcee5f242519.png)

4. Correct.
5. No. I do see the incoming messages.
6. Yes. If I send messages I can see them in the console.

This is my problem: 
1. I have a Python script sending messages. I see them with the device explorer.
2. I have the simple_sample_device script, for listening. But I dont see the Python messages.
3. Using device explorer for sending messages, I see them in the console with simple_sample_device.
4. There is only one registered device.

These are the files: [Github repository](https://github.com/encrypted90/IoTHub-Connector) 

Really thanks for your help

**Comment by pierreca** [comment]
@encrypted90 I think what you're doing in the python script is sending to the device-to-cloud endpoint, which is the endpoint the device uses to send messages to the service (device -> cloud). if you want to receive messages from the service (cloud -> device as opposed to device -> cloud) you have to take your inspiration from the service SDK which sends messages to `/message/devicebound` and has a `to` property in the message that contains the device-specific endpoint for cloud-to-device messages.
we have a bunch of docs here: https://azure.microsoft.com/en-us/documentation/articles/iot-hub-devguide/ that explains all the endpoints and message formats. Specifically for cloud->device, here: https://azure.microsoft.com/en-us/documentation/articles/iot-hub-devguide/#c2d
Do let us know if that helps!

**Comment by encrypted90** [comment]
Yes, I know understand that I should be sending messages to the cloud-to-device endpoint. 

According to the [REST API](https://msdn.microsoft.com/en-us/library/azure/mt590785.aspx), the cloud-to-device endpoint is not available for sending messages with REST. This is correct? 

My script is currently using REST for sending messages.

Thank you! and greetings

**Comment by pierreca** [comment]
@encrypted90 
Yes, the C2D endpoint is AMQP only. We do not provide a python service SDK but we do have node, Java and C# SDKs that you can use, or at least use for inspiration for your python code:
- Node: https://github.com/Azure/azure-iot-sdks/tree/master/node/service
- Java: https://github.com/Azure/azure-iot-sdks/tree/master/java/service
- C#: https://github.com/Azure/azure-iot-sdks/tree/master/csharp/service

I'm going to go ahead and close the issue since we've drifted a little far from the initial issue, feel free to reopen another one if you run into issues with our service SDKs

Thanks!
